File: src/DynamicProgramming/Catalan.java
package DynamicProgramming;

import java.math.BigInteger;

public class Catalan {

    static BigInteger binomCoeff(int n, int k) {
        BigInteger res = new BigInteger("1");
        for (int i = 0; i < k; i++) {
            res = res.multiply(new BigInteger(n - i + ""));
            res = res.divide(new BigInteger(i + 1 + ""));
        }

        return res;
    }

    static BigInteger catalan(int n) {
        BigInteger cat = binomCoeff(2 * n, n);
        cat = cat.divide(new BigInteger(n + 1 + ""));
        return cat;
    }

    public static void main(String[] args) {
        int n = 10;
        System.out.println(catalan(n));
    }
}

File: src/DynamicProgramming/DynamicProgramming/Catalan.py
def binom_coeff(n, k):
    res = 1
    for i in range(k):
        res *= (n - i)
        res //= (i + 1)
    return res


def catalan(n):
    cat = binom_coeff(2 * n, n)
    cat //= (n + 1)
    return cat


if __name__ == "__main__":
    n = 10
    print(catalan(n))